Various improvements

andreasley opened the pull request:
## Improved performance
Previously, all bytes received were passed trough an `map` operation
that is not required, since the bytes are already `UInt8`. This change
improves performance for large transfers considerably.

Improves https://github.com/vapor/vapor/issues/734

## Improved cleanup and safety
- `TCPInternetSocket` and `UDPInternetSocket` now close their descriptors on deinit
- Additional calls to `close()` don't try to close the file descriptor anymore
- Attempts to use a closed `TCPInternetSocket` or `UDPInternetSocket` now throw to prevent using a descriptor that may already be used otherwise

## Improved testing
- `WatchingTests` are now run under Linux, too
- `WatchingTests` should not fail anymore
- Added tests for releasing and using closed sockets

Commit:
Performance improvement

Previously, all bytes received were passed trough an `map` operation
that is not required, since the bytes are already `UInt8`. This change
improves performance for large transfers considerably.

tannernelson commented:
Great find. Tests failed on macOS, though. Re-rerunning to make sure it wasn't a network issue.

andreasley commented:
I've had tests on CI failing for no reason in the last few weeks – tests that succeeded every time when run locally. I think we're pushing the boundaries of Travis with all this TCP stuff. :)
In general, I think it's best to create sockets in tests on any available port instead of a specific one.

tannernelson commented:
Yeah, I'm almost sure it's their issue. The tests were written by the original author of this package so hopefully they don't need to be reworked just for Travis.

LoganWright commented:
I'm not sure why these lines are failing on osx:

https://github.com/vapor/socks/blob/ea09c395312e772ae9ad9a8834c61411fcaec10a/Tests/SocksCoreTests/OptionsTests.swift#L29

https://github.com/vapor/socks/blob/ea09c395312e772ae9ad9a8834c61411fcaec10a/Tests/SocksCoreTests/OptionsTests.swift#L33

Commit:
Added test for repeated watching

`testWatching` might fail randomly because of Dispatch issues. Running
it 1000 times makes (fairly) sure it’s working as expected.

Commit:
Added WatchingTests to XCTestManifest and Linux tests

Commit:
Changed two tests to use TCPInternetSocket instead of a pipe

Hoping that this might fix the failing Travis tests

Commit:
Limiting `close` calls on `TCPInternetSocket`

Calling `close()` multiple times would result in multiple attempts to
close the underlying file descriptor, which shouldn’t happen.

Commit:
TCPInternetSocket calls `close()` on release

When released, a TCPInternetSocket will now attempt to close itself
(and therefore its underlying file descriptor)

Commit:
Made watching test more detailed and reliable

Commit:
UDPSocket calls `close()` on deinit

Commit:
Added lifetime tests

Commit:
Closed TCPInternetSocket and UDPInternetSocket now throw on attempts to use them

Closing a socket means that its descriptor is released and could be
used by other sockets/files/etc.
Attempts to use the descriptor again should be prevented.

Commit:
Fixed a test for Linux
